source/common/config/xds_mux/grpc_mux_impl.h

```diff
@@ -0,0 +1,257 @@
+  PausableAckQueue pausable_ack_queue_;
+
+  // Makes SubscriptionStates, to be held in the subscriptions_ map. Whether this GrpcMux is doing
+  // delta or state of the world xDS is determined by which concrete subclass this variable gets.
+  std::unique_ptr<F> subscription_state_factory_;
+
+  // Map key is type_url.
+  // Only addWatch() should insert into these maps.
+  absl::flat_hash_map<std::string, std::unique_ptr<S>> subscriptions_;
+  absl::flat_hash_map<std::string, std::unique_ptr<WatchMap>> watch_maps_;
+
+  // Determines the order of initial discovery requests. (Assumes that subscriptions are added
+  // to this GrpcMux in the order of Envoy's dependency ordering).
+  std::list<std::string> subscription_ordering_;
+
+  // Whether to enable the optimization of only including the node field in the very first
+  // discovery request in an xDS gRPC stream (really just one: *not* per-type_url).
+  const bool skip_subsequent_node_;
+
+  // State to help with skip_subsequent_node's logic.
+  bool any_request_sent_yet_in_current_stream_{};
+
+  // Used to populate the [Delta]DiscoveryRequest's node field. That field is the same across
+  // all type_urls, and moreover, the 'skip_subsequent_node' logic needs to operate across all
+  // the type_urls. So, while the SubscriptionStates populate every other field of these messages,
+  // this one is up to GrpcMux.
+  const LocalInfo::LocalInfo& local_info_;
+  Common::CallbackHandlePtr dynamic_update_callback_handle_;
+
+  const envoy::config::core::v3::ApiVersion transport_api_version_;
+};
+
+class GrpcMuxDelta : public GrpcMuxImpl<DeltaSubscriptionState, DeltaSubscriptionStateFactory,
+                                        envoy::service::discovery::v3::DeltaDiscoveryRequest,
+                                        envoy::service::discovery::v3::DeltaDiscoveryResponse> {
+public:
+  GrpcMuxDelta(Grpc::RawAsyncClientPtr&& async_client, Event::Dispatcher& dispatcher,
+               const Protobuf::MethodDescriptor& service_method,
+               envoy::config::core::v3::ApiVersion transport_api_version,
+               Random::RandomGenerator& random, Stats::Scope& scope,
+               const RateLimitSettings& rate_limit_settings, const LocalInfo::LocalInfo& local_info,
+               bool skip_subsequent_node);
+
+  // GrpcStreamCallbacks
+  void requestOnDemandUpdate(const std::string& type_url,
+                             const absl::flat_hash_set<std::string>& for_update) override;
+
+  GrpcStream<envoy::service::discovery::v3::DeltaDiscoveryRequest,
+             envoy::service::discovery::v3::DeltaDiscoveryResponse>&
+  grpcStreamForTest() {
+    return grpc_stream_;
+  }
+
+protected:
+  GrpcStreamBase& grpcStream() override { return grpc_stream_; }
+  void sendMessage(envoy::service::discovery::v3::DeltaDiscoveryRequest& msg_proto) override {
+    grpc_stream_.sendMessage(msg_proto);
+  }
+
+private:
+  GrpcStream<envoy::service::discovery::v3::DeltaDiscoveryRequest,
```

Comment:
Can this be a member of `GrpcMuxImpl` (IIUC that class already has `RQ` and `RS` as its template params)?

Reply:
Looks like it should, will give it a try.

Reply:
done.